server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -421,7 +434,11 @@ public IndexRequest source(XContentType xContentType, Object... source) {
      * Sets the document to index in bytes form.
      */
     public IndexRequest source(BytesReference source, XContentType xContentType) {
-        this.source = Objects.requireNonNull(source);
+        assert refCounted.refCount() > 0;
+        if (this.source != null) {
+            this.source.decRef();
```

Comment:
Should we also incRef here on the new source? Since otherwise, if I do:

indexRequest.source(a, type);
indexRequest.source(b, type);

I would have essentially decRef'ed a, which seems counter intuitive?

Reply:
In the line below we currently wrap `a` via

```
 this.source = ReleasableBytesReference.wrap(Objects.requireNonNull(source))
```

so when that gets its reference decremented it's a noop and nothing ever happened to `a` itself. I suppose it would be more efficient to check if `instanceof ReleasableBytesReference` and then `incRef` in that case but since we had no caller that passes a releasable reference I went with this approach for simplicity in the short-run.

Reply:
Thanks for clearing up my confusion here, no need to do any more to this.